File: zatca_integration/overrides/customer.py
"""Customer ZATCA field validation and client mandatory rules."""

from __future__ import annotations

import frappe
from frappe import _

from zatca_integration.common_util import (
    canonicalize_vat_category_field,
    get_customer_vat_category,
    get_vat_validation_settings,
    is_export_vat_category,
    is_vat_number_required,
    validate_ksa_vat_number,
)


def is_overseas_customer(doc) -> bool:
    """Buyer outside KSA, or an Overseas / Export / Deemed Export VAT category."""
    if is_export_vat_category(get_customer_vat_category(doc)):
        return True

    country = str(doc.get("custom_country") or "").strip()
    return bool(country) and country.casefold() != "saudi arabia"


def get_overseas_reason(doc) -> str:
    """
    Why the overseas / export rules apply, as a message fragment.

    The country is checked first, so a Saudi Arabia customer is never described
    as being abroad — for those the Overseas / Export VAT Category is named.
    """
    country = str(doc.get("custom_country") or "").strip()
    if country and country.casefold() != "saudi arabia":
        return _("is outside Saudi Arabia (Country: {0})").format(_(country))

    category = get_customer_vat_category(doc)
    return _('has the overseas / export VAT Category "{0}"').format(_(category) or _("Overseas"))


def is_saudi_registered_company(doc) -> bool:
    """Company in KSA with a domestic (non-Overseas/Export) VAT category."""
    if doc.get("customer_type") != "Company":
        return False
    if str(doc.get("custom_country") or "").strip().casefold() != "saudi arabia":
        return False
    return not is_export_vat_category(get_customer_vat_category(doc))


def has_vat_number(doc) -> bool:
    return bool((doc.get("custom_vat_number") or doc.get("tax_id") or "").strip())


def has_registration(doc) -> bool:
    return bool(
        (doc.get("custom_registration_scheme") or "").strip()
        and (doc.get("custom_registration_number") or "").strip()
    )


def validate(doc, method=None):
    """Server-side Customer rules for ZATCA identification fields."""
    # Keep the VAT Category Select matching its options (invisible characters or
    # a different case make the field render blank and save blank again).
    canonicalize_vat_category_field(doc)

    # Format-check VAT whenever provided for Saudi Arabia customers
    country = (doc.get("custom_country") or "").strip()
    vat_value = (doc.get("custom_vat_number") or "").strip() or (doc.get("tax_id") or "").strip()
    if vat_value and country.casefold() == "saudi arabia":
        validate_ksa_vat_number(vat_value, field_label=_("VAT Number"))

    # Individuals: ZATCA company identification is not required
    if doc.get("customer_type") == "Individual":
        return

    if doc.get("customer_type") != "Company":
        return

    if not country:
        frappe.throw(
            _("Customer Country is required for Company customers."),
            title=_("Missing Country"),
        )

    if is_overseas_customer(doc):
        # VAT number not mandatory for overseas / Export category.
        # Allow save without registration number, but warn so users can fill it.
        registration_number = (doc.get("custom_registration_number") or "").strip()
        if not registration_number:
            frappe.msgprint(
                _(
                    "This company customer {0}. No Registration Number was entered. You can "
                    "still save, but adding a Registration Scheme and Registration Number is "
                    "recommended for ZATCA."
                ).format(get_overseas_reason(doc)),
                title=_("Registration Number Recommended"),
                indicator="orange",
            )
        return

    # Saudi Arabia + Company + domestic VAT category.
    # Whether a VAT Number is mandatory is controlled by "Zatca Settings":
    # - Validate VAT No Against All Company: every company must provide one
    # - Validate VAT No Against Registered Company: only VAT-registered companies
    if (
        is_saudi_registered_company(doc)
        and is_vat_number_required(doc)
        and not (has_vat_number(doc) or has_registration(doc))
    ):
        frappe.throw(
            _(
                "Saudi company customers must have a VAT Number, or both "
                "Registration Scheme and Registration Number."
            ),
            title=_("ZATCA Details Required"),
        )


@frappe.whitelist()
def get_customer_vat_validation_settings():
    """Expose the VAT Number validation switches to the Customer form."""
    return get_vat_validation_settings()

File: zatca_integration/public/js/customer.js
frappe.ui.form.on("Customer", {
	refresh(frm) {
		// Always re-read the switches so a settings change is picked up on next open
		zatca_vat_settings = null;
		toggle_customer_zatca_requirements(frm);
		load_zatca_vat_settings(frm);
		set_vat_number_input_limits(frm);
	},

	before_save(frm) {
		warn_overseas_missing_registration(frm);
	},

	customer_type(frm) {
		toggle_customer_zatca_requirements(frm);
	},

	custom_country(frm) {
		toggle_customer_zatca_requirements(frm);
		warn_overseas_missing_registration(frm);
	},

	tax_category(frm) {
		toggle_customer_zatca_requirements(frm);
		warn_overseas_missing_registration(frm);
		if (!frm.doc.__islocal) {
			frappe.show_alert({
				message: __("Linked addresses will update VAT Category on save."),
				indicator: "blue",
			});
		}
	},

	custom_vat_category(frm) {
		toggle_customer_zatca_requirements(frm);
		warn_overseas_missing_registration(frm);
	},

	custom_vat_number(frm) {
		toggle_customer_zatca_requirements(frm);
		validate_customer_vat_number_input(frm);
	},

	custom_registration_scheme(frm) {
		toggle_customer_zatca_requirements(frm);
	},

	custom_registration_number(frm) {
		toggle_customer_zatca_requirements(frm);
	},
});

function get_vat_category(frm) {
	return (frm.doc.tax_category || frm.doc.custom_vat_category || "").trim();
}

// Keep in sync with zatca_integration/common_util.py:
// - custom_vat_category select: Registered / Unregistered / Overseas / Government / Exempt
// - tax_category link: B2B / B2C / B2G / Export / Non-Resident / Exempt Entity
function normalize_vat_category(value) {
	return String(value || "")
		.trim()
		.replace(/\s+/g, " ")
		.toLowerCase();
}

// VAT-registered buyers (a VAT Number applies when the "registered" switch is on)
const ZATCA_REGISTERED_VAT_CATEGORIES = ["Registered", "B2B", "B2G", "Tax Deductors", "Tax Deductor"];

// Overseas / export-style buyers: no VAT Number, no Saudi national address
const ZATCA_EXPORT_VAT_CATEGORIES = [
	"Overseas",
	"Oversees",
	"Deemed Export",
	"Export / Non-Resident",
	"Export",
];

function is_registered_vat_category(value) {
	const category = normalize_vat_category(value);
	return ZATCA_REGISTERED_VAT_CATEGORIES.some(
		(name) => normalize_vat_category(name) === category
	);
}

function is_export_vat_category(value) {
	const category = normalize_vat_category(value);
	return ZATCA_EXPORT_VAT_CATEGORIES.some((name) => normalize_vat_category(name) === category);
}

function is_overseas(frm) {
	if (is_export_vat_category(get_vat_category(frm))) {
		return true;
	}
	const country = (frm.doc.custom_country || "").trim();
	return !!country && country.toLowerCase() !== "saudi arabia";
}

let zatca_vat_settings = null;
let zatca_vat_settings_request = null;

// Switches live on "Zatca Settings":
// - validate_vat_no_against_all_company: every Company customer needs a VAT No
// - validate_vat_no_against_registered_company: only VAT-registered companies do
function load_zatca_vat_settings(frm) {
	return fetch_zatca_vat_settings().then(() => {
		toggle_customer_zatca_requirements(frm);
		return zatca_vat_settings;
	});
}

function fetch_zatca_vat_settings() {
	if (zatca_vat_settings) {
		return Promise.resolve(zatca_vat_settings);
	}
	if (!zatca_vat_settings_request) {
		zatca_vat_settings_request = frappe
			.xcall("zatca_integration.overrides.customer.get_customer_vat_validation_settings")
			.then((settings) => {
				zatca_vat_settings = settings || {};
				return zatca_vat_settings;
			})
			.catch(() => {
				zatca_vat_settings = {};
				return zatca_vat_settings;
			})
			.then((settings) => {
				zatca_vat_settings_request = null;
				return settings;
			});
	}
	return zatca_vat_settings_request;
}

function is_vat_number_validation_enabled(frm) {
	const settings = zatca_vat_settings || {};
	if (Number(settings.validate_vat_no_against_all_company)) {
		return true;
	}
	if (Number(settings.validate_vat_no_against_registered_company)) {
		return is_registered_vat_category(get_vat_category(frm));
	}
	return false;
}

function is_valid_ksa_vat_number(value) {
	const vat = (value || "").trim();
	if (!vat) {
		return true;
	}
	return /^\d{15}$/.test(vat) && vat.startsWith("3");
}

function set_vat_number_input_limits(frm) {
	if (!frm.fields_dict.custom_vat_number) {
		return;
	}
	frm.set_df_property(
		"custom_vat_number",
		"description",
		__("Exactly 15 digits, starting with 3")
	);
	const $input = frm.get_field("custom_vat_number").$input;
	if ($input && $input.length) {
		$input.attr("maxlength", 15);
	}
}

function validate_customer_vat_number_input(frm) {
	const country = (frm.doc.custom_country || "").trim();
	const vat = (frm.doc.custom_vat_number || "").trim();
	if (!vat || country !== "Saudi Arabia") {
		return;
	}
	if (!is_valid_ksa_vat_number(vat)) {
		frappe.msgprint({
			title: __("Invalid VAT Number"),
			indicator: "red",
			message: __("VAT Number must be exactly 15 digits and start with 3."),
		});
	}
}

// Mirrors zatca_integration.overrides.customer.get_overseas_reason: never tell a
// Saudi Arabia customer that they are abroad — name the VAT Category instead.
function overseas_reason(frm) {
	const country = (frm.doc.custom_country || "").trim();
	if (country && country.toLowerCase() !== "saudi arabia") {
		return __("Country is {0} (outside Saudi Arabia)", [country]);
	}
	return __("VAT Category is {0} (overseas / export)", [
		get_vat_category(frm) || "Overseas",
	]);
}

function warn_overseas_missing_registration(frm) {
	if (frm.doc.customer_type !== "Company" || !is_overseas(frm)) {
		return;
	}
	const registration_number = (frm.doc.custom_registration_number || "").trim();
	if (registration_number) {
		return;
	}
	frappe.show_alert({
		message: __(
			"Registration Number recommended: {0}. You can still save without it.",
			[overseas_reason(frm)]
		),
		indicator: "orange",
	});
}

function toggle_customer_zatca_requirements(frm) {
	const is_company = frm.doc.customer_type === "Company";
	const overseas = is_overseas(frm);

	const has_vat = !!(frm.doc.custom_vat_number || frm.doc.tax_id);
	const has_reg_scheme = !!(frm.doc.custom_registration_scheme || "").trim();
	const has_reg_number = !!(frm.doc.custom_registration_number || "").trim();
	const has_registration = has_reg_scheme && has_reg_number;

	// Individuals: nothing ZATCA-mandatory
	if (!is_company) {
		[
			"custom_country",
			"custom_vat_number",
			"custom_registration_scheme",
			"custom_registration_number",
			"tax_id",
		].forEach((field) => {
			if (frm.fields_dict[field]) {
				frm.toggle_reqd(field, false);
			}
		});
		return;
	}

	// Company: country always required
	if (frm.fields_dict.custom_country) {
		frm.toggle_reqd("custom_country", true);
	}

	if (overseas) {
		// Overseas / Export: VAT not mandatory; registration recommended only (server warns)
		if (frm.fields_dict.custom_vat_number) {
			frm.toggle_reqd("custom_vat_number", false);
		}
		if (frm.fields_dict.custom_registration_scheme) {
			frm.toggle_reqd("custom_registration_scheme", false);
		}
		if (frm.fields_dict.custom_registration_number) {
			frm.toggle_reqd("custom_registration_number", false);
		}
		return;
	}

	// VAT Number is mandatory only when switched on in "Zatca Settings".
	// Wait for the switches before deciding, otherwise leave the current state untouched.
	if (!zatca_vat_settings) {
		load_zatca_vat_settings(frm);
		return;
	}

	if (is_vat_number_validation_enabled(frm)) {
		// Need VAT OR (scheme + number). Require whichever side is still missing.
		if (frm.fields_dict.custom_vat_number) {
			frm.toggle_reqd("custom_vat_number", !has_registration);
		}
		if (frm.fields_dict.custom_registration_scheme) {
			frm.toggle_reqd("custom_registration_scheme", !has_vat);
		}
		if (frm.fields_dict.custom_registration_number) {
			frm.toggle_reqd("custom_registration_number", !has_vat && has_reg_scheme);
		}
		return;
	}

	// Validation switched off: VAT Number / registration details stay optional
	["custom_vat_number", "custom_registration_scheme", "custom_registration_number"].forEach(
		(field) => {
			if (frm.fields_dict[field]) {
				frm.toggle_reqd(field, false);
			}
		}
	);
}
